File: 0402-remove-k-digits/0402-remove-k-digits.java
class Solution {
    public String removeKdigits(String s, int k) {
        int N = s.length();
        
        if (k == N) {
            return "0";
        }
        
        Stack<Character> stk = new Stack<>();
        
        for (int i = 0; i < N; i++) {
            
            while (!stk.isEmpty() && k > 0 && stk.peek() > s.charAt(i)) {
                stk.pop();
                k--;
            }
            
            stk.push(s.charAt(i));
        }
        
        while (!stk.isEmpty() && k > 0) {
            stk.pop();
            k--;
        }
        
        char[] res = new char[stk.size()];
        
        for (int i = res.length - 1; i >= 0; i--) {
            res[i] = stk.pop();
        }
        
        k = 0;
        
        for (int i = 0; i < res.length; i++) {
            if (res[i] == '0') {
                k++;
            } else {
                break;
            }
        }
        
        if (k == res.length) {
            return "0";
        }
        
        return new String(res, k, res.length - k);
    }
}

File: 0402-remove-k-digits/0402-remove-k-digits.cpp
class Solution {
public:
    string removeKdigits(string num, int k) {
        int n = num.size();
        
        if (n == k) {
            return "0";
        }
        
        int i = 0;
        stack<int> st;
        
        while (i < n && k > 0) {
            if (st.empty() || st.top() <= num[i]) {
                st.push(num[i]);
            } else {
                while (!st.empty() && k > 0 && st.top() > num[i]) {
                    k--;
                    st.pop();
                }
                st.push(num[i]);
            }
            i++;
        }
        
        while (i < n) {
            st.push(num[i]);
            i++;
        }
        
        while (k > 0) {
            st.pop();
            k--;
        }
        
        k = st.size();
        vector<int> vt(k);
        
        while (!st.empty()) {
            vt[--k] = st.top(); st.pop();
        }
        
        int y = 0;
        while (y < vt.size() && vt[y] == '0') {
            y++;
        }
        
        string s(vt.begin() + y, vt.end());
        
        return s == "" ? "0" : s;
    }
};